syz-verifier/main.go

```diff
@@ -0,0 +1,329 @@
+	for idx := 1; idx < len(kis); idx++ {
+		cfg := kis[idx].Cfg
+
+		// TODO: pass the configurations that should be the same for all kernels in a default config file in order to avoid this checks and add testing
+		if workdir != cfg.Workdir {
+			log.Fatalf("working directory mismatch")
+		}
+		if target != cfg.Target {
+			log.Fatalf("target mismatch")
+		}
+		if sysTarget != cfg.SysTarget {
+			log.Fatalf("system target mismatch")
+		}
+		if addr != kis[idx].Cfg.RPC {
+			log.Fatalf("tcp address mismatch")
+		}
+	}
+
+	exe := sysTarget.ExeExtension
+	runnerBin := filepath.Join(cfg.Syzkaller, "bin", target.OS+"_"+target.Arch, "syz-runner"+exe)
+	if !osutil.IsExist(runnerBin) {
+		log.Fatalf("bad syzkaller config: can't find %v", runnerBin)
+	}
+	execBin := cfg.ExecutorBin
+	if !osutil.IsExist(execBin) {
+		log.Fatalf("bad syzkaller config: can't find %v", execBin)
+	}
+
+	crashdir := filepath.Join(workdir, "crashes")
+	osutil.MkdirAll(crashdir)
+	for idx := range kis {
+		OS, Arch := target.OS, target.Arch
+		targetPath := OS + "-" + Arch + "-" + strconv.Itoa(idx)
+		osutil.MkdirAll(filepath.Join(workdir, targetPath))
+		osutil.MkdirAll(filepath.Join(crashdir, targetPath))
+	}
+
+	for idx, ki := range kis {
+		var err error
+		ki.Reporter, err = report.NewReporter(ki.Cfg)
+		if err != nil {
+			log.Fatalf("failed to create reporter for instance-%d: %v", idx, err)
+		}
+	}
+
+	calls := make(map[*prog.Syscall]bool)
+	for _, id := range cfg.Syscalls {
+		calls[target.Syscalls[id]] = true
+	}
+
+	vrf := &verf.Verifier{
+		Crashdir:    crashdir,
+		Kis:         kis,
+		Target:      target,
+		ChoiceTable: target.BuildChoiceTable(nil, calls),
+		Rnd:         rand.New(rand.NewSource(time.Now().UnixNano() + 1e12)),
+		RunnerBin:   runnerBin,
+		ExecutorBin: execBin,
+		Addr:        addr,
+	}
+
+	srv, err := startRPCServer(vrf)
+	if err != nil {
+		log.Fatalf("failed to initialise RPC server: %v", err)
+	}
+
+	for idx, ki := range kis {
+		go func(ki *verf.KernelInfo, idx int) {
+			// TODO: consider supportin more VMs per Pool
+			inst, err := ki.VmPool.Create(0)
+			if err != nil {
+				log.Fatalf("failed to create instance: %v", err)
+			}
+			defer inst.Close()
+
+			fwdAddr, err := inst.Forward(srv.port)
+			if err != nil {
+				log.Fatalf("failed to set up port forwarding: %v", err)
+			}
+
+			runnerBin, err := inst.Copy(vrf.RunnerBin)
+			if err != nil {
+				log.Fatalf(" failed to copy runner binary: %v", err)
+			}
+			_, err = inst.Copy(vrf.ExecutorBin)
+			if err != nil {
+				log.Fatalf("failed to copy executor binary: %v", err)
+			}
+
+			cmd := instance.RunnerCmd(runnerBin, fwdAddr, vrf.Target.OS, vrf.Target.Arch, idx)
+			outc, errc, err := inst.Run(ki.Cfg.Timeouts.VMRunningTime, vrf.VmStop, cmd)
+			if err != nil {
+				log.Fatalf("failed to start runner: %v", err)
+			}
+
+			// TODO: implement VM restart functionality and record the return value of inst.MonitorExecution
```

Comment:
Will a `for {}` loop around this function do?
We need some way to restart them. Currently if even 1 VM stops/crashes or ki.Cfg.Timeouts.VMRunningTime elapses, the process will hang.

Reply:
I think so, it's a quick solution for now